hepstats/hypotests/toyutils.py

```diff
@@ -0,0 +1,356 @@
+        except KeyError:
+            return 0
+
+    def generate_and_fit_toys(self, ntoys, poigen, poieval, printfreq=0.2):
+        """
+        Generate and fit toys for at a given POI (poigen). The toys are then fitted, and the likelihood
+        is profiled at the values of poigen and poieval.
+
+        Args:
+            ntoys (int): number of toys to generate
+            poigen (POI): POI used to generate the toys
+            poieval (POIarray, optional): POI values to evaluate the loss function
+            printfreq: print frequency of the toys generation
+        """
+
+        minimizer = self.minimizer
+        param = poigen.parameter
+
+        toys_loss = self.toys_loss(poigen.name)
+        sampler = toys_loss.data
+
+        bestfit = np.empty(ntoys)
+        nll_bestfit = np.empty(ntoys)
+        nlls = {p: np.empty(ntoys) for p in poieval}
+
+        printfreq = ntoys * printfreq
+
+        samples = self.sample(sampler, int(ntoys*1.2), poigen)
+
+        try:
+            toysresult = self.get_toyresult(poigen, poieval)
+            poieval = toysresult.poieval
+        except KeyError:
+            toysresult = ToyResult(poigen, poieval)
+            self.add_toyresult(toysresult)
+
+        for i in range(ntoys):
+            converged = False
+            toprint = i % printfreq == 0
+            while converged is False:
+                try:
+                    next(samples)
+                except StopIteration:
+                    to_gen = ntoys - i
+                    samples = self.sample(sampler, int(to_gen*1.2), poigen)
+                    next(samples)
+
+                minimum = minimizer.minimize(loss=toys_loss)
+                converged = minimum.converged
+
+                if not converged:
+                    self.set_dependents_to_bestfit()
+                    continue
+
+                bestfit[i] = minimum.params[param]["value"]
+                nll_bestfit[i] = pll(minimizer, toys_loss, POI(param, bestfit[i]))
+
+                for p in poieval:
+                    nlls[p][i] = pll(minimizer, toys_loss, p)
+
+            if toprint:
+                print("{0} toys generated, fitted and scanned!".format(i))
+
+            if i > ntoys:
+                break
+            i += 1
+
+        if (poigen, poieval) not in self.keys():
+            toysresult = ToyResult(p, poieval)
+            self.add_toyresult(toysresult)
+        else:
+            toysresult = self.get_toyresult(poigen, poieval)
+
+        toysresult.add_entries(bestfit=bestfit, nll_bestfit=nll_bestfit, nlls=nlls)
+
+    def keys(self):
+        return self._toys.keys()
+
+    def values(self):
+        return self._toys.values()
+
+    def toyresults_to_dict(self):
+        """
+        Returns a list of all the toy results converted into dictionnaries
+        """
+        return [v.to_dict() for v in self.values()]
+
+    def to_yaml(self, filename):
+        """
+        Save the toys into a yaml file.
```

Comment:
You need to specify that the result is stored under key `toys`. This is relevant for the method `toysresults_from_yaml` below, where the key is immediately requested. BTW, there one should probably try/except for the presence of the relevant key, and throw a "FormatError" or alike in case the YAML file is not in the expected format.
Just trying tobe user-friendly ...